lib/ssh/src/ssh_acceptor_subsup.erl

```diff
@@ -0,0 +1,142 @@
+%%
+%% %CopyrightBegin%
+%%
+%% Copyright Ericsson AB 2008-2024. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%
+%% %CopyrightEnd%
+%%
+
+%%
+%%----------------------------------------------------------------------
+%% Purpose: The acceptor supervisor for ssh servers hangs under 
+%%          ssh_system_sup.
+%%----------------------------------------------------------------------
+
+-module(ssh_acceptor_subsup).
+-moduledoc false.
+-behaviour(supervisor).
+
+-include("ssh.hrl").
+
+-export([start_link/3,
+         listen/2
+        ]).
+
+%% Supervisor callback
+-export([init/1]).
+
+%%%=========================================================================
+%%%  API
+%%%=========================================================================
+start_link(SystemSup, Address, Options) ->
+    case supervisor:start_link(?MODULE, [SystemSup, Address, Options]) of
+        {error, {shutdown, {failed_to_start_child, _, Error}}} ->
+            {error,Error};
+        Other ->
+            Other
+    end.
+
+%%%=========================================================================
+%%%  Supervisor callback
+%%%=========================================================================
+init([SystemSup, Address=#address{port=Port}, Options]) ->
+    ssh_lib:set_label(server, acceptor_subsup),
+    %% Initial start of ssh_acceptor_sup for this port
+    Self = self(),
+    NumAcceptors = case ?GET_OPT(parallel_login, Options) of
+                       true -> 1;
+                       false -> 1;
+                       N -> N
+                   end,
+    {Sock, Opts} = case ?GET_INTERNAL_OPT(lsocket, Options, undefined) of
+        {LSock, SockOwner} ->
+            %% A listening socket (or fd option) was provided in the ssh:daemon call
+            case inet:sockname(LSock) of
+                {ok, {_, Port}} ->
+                    %% A usable, open LSock
+                    spawn_link(fun() ->
```

Comment:
I realize this is probably the fault of the original design, but transferring ownership of the listen socket from ssh user process context feels like a bad idea. Why not make the listen call in a dedicated acceptor process? In current solution that would be in the ssh_acceptor_subsup process.

Reply:
I was thinking the same thing, but see here: https://github.com/erlang/otp/blob/9859b7281e447681a38da9f34292a7825f437de7/lib/ssh/src/ssh.erl#L599-L613

Reply:
Hum, well maybe we need it then. I am still not fond of making an explicit receive in the users process like that without more control like in gen_server:call for instance.

Reply:
I'm not a huge fan of that, either... Maybe a greater refactoring attempt is in order then, but I think that should involve more planning.

Reply:
Maybe we could go with a "good enough bug fix" and then make a refactor later in new master after OTP-27 (as it is really soon)

Reply:
I think that we should be able to move away from the above approach. Th concept of fd is provided is inherited from httpd_sup and  we should be able to handle it like that, that is port is also specified but not opened as it is preopend.
And the case of 0 is the port is only for testing and I think we should rather report the the chosen port bac to the caller
then the way it is done now.